include/sampleflow/consumers/autocovariance_dim_1.h

```diff
@@ -0,0 +1,255 @@
+#define SAMPLEFLOW_CONSUMERS_AUTOCOVARIANCEDIM1_H
+
+#include <sampleflow/consumer.h>
+#include <sampleflow/types.h>
+#include <mutex>
+
+#include <boost/numeric/ublas/matrix.hpp>
+
+
+namespace SampleFlow
+{
+  namespace Consumers
+  {
+    /**
+     * A Consumer class that implements computing the running sample autocovariance function:
+     * \hat\gamma(k)=\frac{1}{n}\sum_{t=1}^{n-k}{(x_{t+k}-\bar{x})(x_{t}-\bar{x})}
+     *
+     * ### Threading model ###
+     *
+     * The implementation of this class is thread-safe, i.e., its
+     * consume() member function can be called concurrently and from multiple
+     * threads.
+     *
+     *
+     * @tparam InputType The C++ type used for the samples $x_k$. In
+     *   order to compute covariances, the same kind of requirements
+     *   have to hold as listed for the Covariance class.
+     */
+    template <typename InputType>
+    class autocovariancedim1 : public Consumer<InputType>
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+        using scalar_type = InputType;
+        /**
+         * The type of the information generated by this class, i.e., in which
+         * the autocovariance function is computed. By itself, for autocovariance there should
+         * be enough to use vector. However, because we keep updating autocovariance, we need to save
+         * parts used in calculation
+         */
+
+        using value_type = boost::numeric::ublas::matrix<scalar_type>;
+
+        /**
+         * Constructor.
+         */
+
+        autocovariancedim1();
+
+        /**
+         * Process one sample by updating the previously computed covariance
+         * matrix using this one sample.
+         *
+         * @param[in] sample The sample to process.
+         * @param[in] aux_data Auxiliary data about this sample. The current
+         *   class does not know what to do with any such data and consequently
+         *   simply ignores it.
+         */
+        virtual
+        void
+        consume (InputType     sample,
+                 AuxiliaryData aux_data) override;
+
+        /**
+         * A function that returns the covariance matrix computed from the
+         * samples seen so far. If no samples have been processed so far, then
+         * a default-constructed object of type InputType will be returned.
+         *
+         * @return The computed covariance matrix.
+         */
+        value_type
+        get () const;
+
+      private:
+        /**
+         * A mutex used to lock access to all member variables when running
+         * on multiple threads.
+         */
+
+        mutable std::mutex mutex;
+
+        /**
+         * The current value of $\bar x_k$ as described in the introduction
+         * of this class.
+         */
+
+        InputType           current_mean;
+
+        /**
+         * Currently describes lag of autocovariance
+         */
+
+        InputType 			k;
```

Comment:
I suspect that you probably want to move this variable to a different place, but in any case, the data type needs to be an integer type, not whatever the type of the samples is. I would choose `unsigned int`, because you describe a lag or length, and that's a positive integer.